Add a test suite for continuous integration

tmattio opened the pull request:
This PR adds a test suite (for now only testing that we can lock and build mdx, irmin and rwo) and a GitHub Action workflow to run the tests on different compiler versions.

A better CI setup would be to try to build all the revdeps of the published packages (like opam-repo CI does), but that's a larger effort, so this PR aims at giving us some stability assertions for known opam-monorepo users in the immediate term.

@hannesm are there Mirage projects that you'd like to add to the test suite to make sure that publication on `opam-overlays` don't break them?

cc @avsm

tmattio commented:
I don't think there's a way to run GA workflows that haven't been committed on the main branch, so if/when this looks reasonable, we can merge and I'll fix the eventual CI failures.

Leonidas-from-XIV commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,40 @@
+      matrix:
+        ocaml-compiler:
+          - 4.12.x
+          - 4.13.x
```
I think we can skip `4.13`. I'd suggest oldest supported version, newest 4.x version, newest 5.x version for the selections.

Leonidas-from-XIV commented on test/mdx.t/run.t:
```diff
@@ -0,0 +1,14 @@
+  ==> Using lockfile $TESTCASE_ROOT/opam-overlays-test-mdx.opam.locked
+  Successfully pulled 18/18 repositories
+  $ dune build
+  $ dune exec ocaml-mdx -- --version
```
Should probably make sure that the `mdx` version here matches what the lockfile said it is supposed to be, otherwise releasing `mdx.2.2.1` will immediately make the testsuite fail.

Leonidas-from-XIV reviewed (changes requested):
Thanks for the testsuite! I think this is a good start and we should merge this ASAP to have something and can iterate on it later on.

I've added a few comments to make it less fragile. The only thing worse than missing CI is CI that fails for unrelated reasons, so we should make sure to make it a bit more resilient without sacrificing testing the functionality.

tmattio commented:
Thanks for the review @Leonidas-from-XIV.

I've addressed most of your comments. For the CLI versions, I suggest we just run the binaries and see if we get error code 0. We don't actually need to verify the version, just making sure the build was successful by running the binary is enough I think.
